lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -245,13 +245,57 @@ LogicalResult PipelineStageLowering::matchAndRewrite(
   return success();
 }
 
+namespace {
+/// Eliminate back-to-back wrap-unwraps to reduce the number of ESI channels.
+struct RemoveWrapUnwrap : public OpConversionPattern<WrapValidReady> {
+public:
+  RemoveWrapUnwrap(MLIRContext *ctxt) : OpConversionPattern(ctxt) {}
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(WrapValidReady wrap, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const final;
+};
+} // anonymous namespace
+
+LogicalResult
+RemoveWrapUnwrap::matchAndRewrite(WrapValidReady wrap,
+                                  ArrayRef<Value> stageOperands,
+                                  ConversionPatternRewriter &rewriter) const {
+  auto wrapUses = wrap.chanOutput().getUseList();
+  // This pass requires that implicit forks already be lowered by the
+  // abstract-to-physical pass. Meaning exactly one user.
+  if (!wrapUses->hasOneUse())
```

Comment:
I don't know if it's any better, but instead of the use list of `wrap.chanOutput()`, I think you can just use helpers on it directly. Like `wrap.chanOutput().hasOneUse()` and `wrap.chanOutput.use_begin().getOwner()`.